src/Nethermind/Nethermind.Core/Caching/NonBlockingLruCache.cs

```diff
@@ -0,0 +1,266 @@
+                return true;
+            }
+
+            return false;
+        }
+
+        public bool Contains(TKey key)
+        {
+            // Use TryGet to schedule the node for Lru.
+            return TryGet(key, out _);
+        }
+
+        private void ScheduleLruAccounting(LinkedListNode<LruCacheItem> node, bool isGet = true)
+        {
+            _accesses.Enqueue(node);
+
+            long count = Interlocked.Increment(ref _toProcess);
+            if (count < _maxCapacity && (isGet || _cacheMap.Count < _maxCapacity))
+            {
+                // Can wait
+                return;
+            }
+
+            // Set working if it wasn't (via atomic Interlocked).
+            if (Interlocked.CompareExchange(ref _processingLru, 1, 0) == 0)
+            {
+                Volatile.Write(ref _toProcess, 0);
+                // Wasn't working, schedule.
+                ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+            }
+        }
+
+        void IThreadPoolWorkItem.Execute()
+        {
+            try
+            {
+                ProcessExpiredItems();
+            }
+            catch
+            {
+                // Tear it all down
+                Clear();
+                _processingLru = 0;
+            }
+        }
+
+        private void ProcessExpiredItems()
+        {
+            CancellationToken ct = _cts.Token;
+            while (true)
+            {
+                while (!ct.IsCancellationRequested && _accesses.TryDequeue(out LinkedListNode<LruCacheItem>? node))
+                {
+                    bool exists = _cacheMap.ContainsKey(node.Value.Key);
+
+                    ref LinkedListNode<LruCacheItem>? leastRecentlyUsed = ref _leastRecentlyUsed;
+                    if (leastRecentlyUsed is null)
+                    {
+                        // Clear has been called
+                        break;
+                    }
+
+                    if (!exists)
+                    {
+                        LinkedListNode<LruCacheItem>.Remove(ref leastRecentlyUsed, node);
+                    }
+                    else if (node.Next is null)
+                    {
+                        LinkedListNode<LruCacheItem>.AddMostRecent(ref leastRecentlyUsed, node);
+                    }
+                    else
+                    {
+                        LinkedListNode<LruCacheItem>.MoveToMostRecent(ref leastRecentlyUsed, node);
+                    }
+                }
+
+                while (!ct.IsCancellationRequested && _cacheMap.Count > _maxCapacity)
+                {
+                    LinkedListNode<LruCacheItem>? leastRecentlyUsed = _leastRecentlyUsed;
+                    if (leastRecentlyUsed is null)
+                    {
+                        // Clear has been called
+                        break;
+                    }
+
+                    _cacheMap.Remove(leastRecentlyUsed.Value.Key, out leastRecentlyUsed);
```

Comment:
I see that `.Remove` is called in two places only, one on `Delete` and second in here.